// validate-india/dist/root.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var ocr_1 = require("./ocr");
/**
 * Base class to validate documents IDs
 *
 * @export
 * @class Root
 */
var Root = /** @class */ (function () {
    function Root(regex, cleanRegex) {
        this.regex = new RegExp(regex);
        if (cleanRegex) {
            this.cleanRegex = cleanRegex;
        }
    }
    Root.prototype.isValid = function (value) {
        if (typeof value === 'number') {
            value = value.toString();
        }
        return this.regex.test(value);
    };
    Root.prototype.clean = function (value) {
        if (!value) {
            return null;
        }
        if (this.cleanRegex) {
            return value.replace(this.cleanRegex, '');
        }
        else {
            return value;
        }
    };
    return Root;
}());
exports.Root = Root;
/**
 * Class to extract Document IDs from image using Google Vision API
 *
 * @export
 * @class OcrRoot
 * @extends {Root}
 */
var OcrRoot = /** @class */ (function (_super) {
    __extends(OcrRoot, _super);
    function OcrRoot() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Retrieves Document number from an image url provided
     *
     * @param {string} uri The image URL (should be publicly accessible)
     * @param {boolean} raw Whether to return the raw value extracted from the image instead of cleaning the data
     * @returns {(Promise<string | null>)} Document number
     * @memberof OcrRoot
     */
    OcrRoot.prototype.extractFromImage = function (uri, raw) {
        if (raw === void 0) { raw = false; }
        return __awaiter(this, void 0, void 0, function () {
            var text, textTokens, result;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, ocr_1.extractTextFromImage(uri)];
                    case 1:
                        text = _a.sent();
                        textTokens = text.split('\n');
                        result = null;
                        textTokens.forEach(function (token) {
                            if (_this.regex.test(token)) {
                                result = token;
                                return;
                            }
                        });
                        return [2 /*return*/, this.clean(result)];
                }
            });
        });
    };
    return OcrRoot;
}(Root));
exports.OcrRoot = OcrRoot;
